File: Odysseus and the Bag of Winds/claude1.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
const ll INF = LLONG_MAX / 2;

int main(){
    int T;
    scanf("%d", &T);
    while(T--){
        int M, A;
        scanf("%d %d", &M, &A);
        int Aidx = A - 1;

        vector<vector<int>> C(M, vector<int>(M));
        for(int i = 0; i < M; i++)
            for(int j = 0; j < M; j++)
                scanf("%d", &C[i][j]);

        int full = 1 << M;
        vector<vector<array<ll,2>>> dp(full, vector<array<ll,2>>(M, {INF, INF}));
        dp[1][0][0] = 0;

        for(int mask = 1; mask < full; mask++){
            if(!(mask & 1)) continue;
            bool aReached = (mask & (1 << Aidx)) != 0;

            for(int last = 0; last < M; last++){
                if(!(mask & (1 << last))) continue;

                for(int bag = 0; bag < 2; bag++){
                    ll cur = dp[mask][last][bag];
                    if(cur >= INF) continue;

                    for(int j = 0; j < M; j++){
                        if(mask & (1 << j)) continue;
                        int nmask = mask | (1 << j);

                        ll nc = cur + C[last][j];
                        if(nc < dp[nmask][j][bag])
                            dp[nmask][j][bag] = nc;

                        if(bag == 0 && aReached){
                            ll hc = cur + C[last][j] / 2;
                            if(hc < dp[nmask][j][1])
                                dp[nmask][j][1] = hc;
                        }
                    }
                }
            }
        }

        int fullmask = full - 1;
        ll ans = min(dp[fullmask][M-1][0], dp[fullmask][M-1][1]);
        printf("%lld\n", ans);
    }
    return 0;
}

File: Odysseus and the Bag of Winds/claude1.py
import sys

def main():
    data = sys.stdin.read().split()
    idx = 0
    t = int(data[idx]); idx += 1
    
    out = []
    INF = float('inf')
    
    for _ in range(t):
        m = int(data[idx]); idx += 1
        a = int(data[idx]); idx += 1
        a_idx = a - 1
        
        c = []
        for i in range(m):
            row = data[idx:idx+m]
            idx += m
            c.append([int(x) for x in row])
        
        full = 1 << m
        dp = [[[INF, INF] for _ in range(m)] for _ in range(full)]
        dp[1][0][0] = 0
        
        for mask in range(1, full):
            if not (mask & 1):
                continue
            a_reached = (mask & (1 << a_idx)) != 0
            
            for last in range(m):
                if not (mask & (1 << last)):
                    continue
                for bag in range(2):
                    cur = dp[mask][last][bag]
                    if cur == INF:
                        continue
                    for j in range(m):
                        if mask & (1 << j):
                            continue
                        nmask = mask | (1 << j)
                        
                        nc = cur + c[last][j]
                        if nc < dp[nmask][j][bag]:
                            dp[nmask][j][bag] = nc
                        
                        if bag == 0 and a_reached:
                            hc = cur + c[last][j] // 2
                            if hc < dp[nmask][j][1]:
                                dp[nmask][j][1] = hc
        
        fullmask = full - 1
        ans = min(dp[fullmask][m-1][0], dp[fullmask][m-1][1])
        out.append(str(ans))
    
    print("\n".join(out))

main()